c++/src/kj/compat/http.c++

```diff
@@ -4093,6 +4093,103 @@ private:
   kj::ArrayPtr<byte> leftover;
 };
 
+class AsyncIoStreamWithDeferredInput final: public kj::AsyncIoStream {
+  // This AsyncIoStream immediately allows write to the given wrapped AsyncIoStream
+  // but will delay reads until after a guard promise is resolved. When the guard
+  // promise resolves, it may provide a released buffer that will be read out first.
+  // The primary use case for this impl is to support pipelined CONNECT calls which
+  // optimistically allow outbound writes to happen while establishing the CONNECT
+  // tunnel has not yet been completed. If the guard promise rejects, the stream
+  // is permanently errored and existing pending calls (reads and writes) are canceled.
+public:
+  AsyncIoStreamWithDeferredInput(
+      kj::Own<kj::AsyncIoStream> inner,
+      kj::Promise<HttpInputStreamImpl::ReleasedBuffer> guard)
+      : inner(kj::mv(inner)),
+        guard(handleGuard(kj::mv(guard))) {}
+
+  // AsyncInputStream
+  Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      return kj::cp(*exception);
+    }
+    return guard.addBranch().then([this, buffer, minBytes, maxBytes] {
+      return inner->tryRead(buffer, minBytes, maxBytes);
+    });
+  }
+
+  Maybe<uint64_t> tryGetLength() override {
+    return nullptr;
+  }
+
+  kj::Promise<uint64_t> pumpTo(AsyncOutputStream& output, uint64_t amount) override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      return kj::cp(*exception);
+    }
+    return guard.addBranch().then([this, &output, amount] {
+      return inner->pumpTo(output, amount);
+    });
+  }
+
+  // AsyncOutputStream
+
+  void shutdownWrite() override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      throw *exception;
+    }
+    inner->shutdownWrite();
+  }
+
+  kj::Maybe<kj::Promise<uint64_t>> tryPumpFrom(AsyncInputStream& input, uint64_t amount) override {
+    KJ_IF_MAYBE(exception, permanentException) {
```

Comment:
There's a lot of effort and bookkeeping overhead introduced here to make sure that writes end up propagating the status exception. I'm not really sure it's needed. You could say it's the app's responsibility to wait on the status and cancel any writes if it fails. This kind of bookkeeping has a cost, it's the kind of thing I've been removing from the RPC system in order to make capnp RPC faster.

Reply:
Yeah, I was debating that. The reason I went with this approach was to actually keep the app from having to go through the trouble of doing this book keeping since it is likely just as much overhead and trouble there. That said, I'm definitely not that happy with this as is.